Firestore/core/src/firebase/firestore/remote/stream_objc_bridge.mm

```diff
@@ -16,17 +16,158 @@
+    }
+    return data;
+  }
+}
+
+std::string ToHexString(const grpc::ByteBuffer& buffer) {
+  std::vector<grpc::Slice> slices;
+  grpc::Status status = buffer.Dump(&slices);
+
+  std::stringstream output;
+  // The output will look like "0x00 0x0a"
+  output << std::hex << std::setfill('0') << std::setw(2);
+  for (const auto& slice : slices) {
+    for (uint8_t c : slice) {
+      output << "0x" << static_cast<int>(c) << " ";
+    }
+  }
+
+  return output.str();
+}
+
+template <typename Proto>
+Proto* ToProto(const grpc::ByteBuffer& message, Status* out_status) {
+  NSError* error = nil;
+  Proto* proto = [Proto parseFromData:ToNsData(message) error:&error];
+  if (!error) {
+    *out_status = Status::OK();
+    return proto;
+  }
+
+  std::string error_description = StringFormat(
+      "Unable to parse response from the server.\n"
+      "Underlying error: %s\n"
+      "Expected class: %s\n"
+      "Received value: %s\n",
+      error, [Proto class], ToHexString(message));
+
+  *out_status = {FirestoreErrorCode::Internal, error_description};
+  return nil;
+}
+
+grpc::ByteBuffer ConvertToByteBuffer(NSData* data) {
+  grpc::Slice slice{[data bytes], [data length]};
+  return grpc::ByteBuffer{&slice, 1};
+}
+
+}  // namespace
+
 bool IsLoggingEnabled() {
   return [FIRFirestore isLoggingEnabled];
 }
 
+GCFSListenRequest* WatchStreamSerializer::CreateWatchRequest(
+    FSTQueryData* query) const {
+  GCFSListenRequest* request = [GCFSListenRequest message];
+  request.database = [serializer_ encodedDatabaseID];
+  request.addTarget = [serializer_ encodedTarget:query];
+  request.labels = [serializer_ encodedListenRequestLabelsForQueryData:query];
+  return request;
+}
+
+GCFSListenRequest* WatchStreamSerializer::CreateUnwatchRequest(
+    TargetId target_id) const {
+  GCFSListenRequest* request = [GCFSListenRequest message];
+  request.database = [serializer_ encodedDatabaseID];
+  request.removeTarget = target_id;
+  return request;
+}
+
+grpc::ByteBuffer WatchStreamSerializer::ToByteBuffer(
+    GCFSListenRequest* request) const {
+  return ConvertToByteBuffer([request data]);
+}
+
+NSString* WatchStreamSerializer::Describe(GCFSListenRequest* request) const {
+  return [request description];
+}
+
+NSString* WatchStreamSerializer::Describe(GCFSListenResponse* response) const {
+  return [response description];
+}
+
+FSTWatchChange* WatchStreamSerializer::ToWatchChange(
+    GCFSListenResponse* proto) const {
+  return [serializer_ decodedWatchChange:proto];
+}
+
+SnapshotVersion WatchStreamSerializer::ToSnapshotVersion(
+    GCFSListenResponse* proto) const {
+  return [serializer_ versionFromListenResponse:proto];
+}
+
+GCFSListenResponse* WatchStreamSerializer::ParseResponse(
+    const grpc::ByteBuffer& message, Status* out_status) const {
+  return ToProto<GCFSListenResponse>(message, out_status);
+}
+
+void WatchStreamDelegate::NotifyDelegateOnOpen() {
+  id<FSTWatchStreamDelegate> delegate = delegate_;
```

Comment:
can we make delegate_ `id<FSTWatchStreamDelegate>` instead of `id` and call on `delegate_` directly?

Reply:
In the Objective-C implementation the delegate is stored in the FSTStream base class, so there it's not known which subtype is in effect. Here there's a watch-specific delegate wrapper so it could be typed more strongly.

Reply:
Done.